ml-proto/spec/int.ml

```diff
@@ -0,0 +1,197 @@
+(* WebAssembly-compatible i32 implementation *)
+
+module type RepresentationType =
+sig
+  type t  
+  val add : t -> t -> t
+  val min_int : t
+  val zero : t
+  val one : t
+  val minus_one : t                        
+  val shift_left : t -> int -> t
+  val shift_right : t -> int -> t
+  val logand : t -> t -> t
+  val lognot : t -> t
+  val logor : t -> t -> t
+  val logxor : t -> t -> t                    
+  val sub : t -> t -> t
+  val div : t -> t -> t
+  val mul : t -> t -> t
+  val rem : t -> t -> t                        
+  val shift_right_logical : t -> int -> t
+  val of_int : int -> t
+  val to_int : t -> int                        
+  val of_string : string -> t
+  val to_string : t -> string
+
+  val bitwidth : int 
+end
+
+module type S =
+sig
+  type t
+  type bits 
+
+  val of_bits : bits -> t
+  val to_bits : t -> bits
+
+  val zero : t
+
+  val add : t -> t -> t
+  val sub : t -> t -> t
+  val mul : t -> t -> t
+  val div_s : t -> t -> t
+  val div_u : t -> t -> t
+  val rem_s : t -> t -> t
+  val rem_u : t -> t -> t
+  val and_ : t -> t -> t
+  val or_ : t -> t -> t
+  val xor : t -> t -> t
+  val shl : t -> t -> t
+  val shr_s : t -> t -> t
+  val shr_u : t -> t -> t
+  val clz : t -> t
+  val ctz : t -> t
+  val popcnt : t -> t
+  val eq : t -> t -> bool
+  val ne : t -> t -> bool
+  val lt_s : t -> t -> bool
+  val lt_u : t -> t -> bool
+  val le_s : t -> t -> bool
+  val le_u : t -> t -> bool
+  val gt_s : t -> t -> bool
+  val gt_u : t -> t -> bool
+  val ge_s : t -> t -> bool
+  val ge_u : t -> t -> bool
+
+  val of_int : int -> t
+  val of_string : string -> t
+  val to_string : t -> string
+end
+    
+module Make(Rep : RepresentationType) : S with type bits = Rep.t and type t = Rep.t =
+struct
+  (*
+   * Unsigned comparison in terms of signed comparison.
+   *)
+  let cmp_u x op y =
+    op (Rep.add x Rep.min_int) (Rep.add y Rep.min_int)
+
+  (*
+   * Unsigned division and remainder in terms of signed division; algorithm from
+   * Hacker's Delight, Second Edition, by Henry S. Warren, Jr., section 9-3
+   * "Unsigned Short Division from Signed Division".
+   *)
+  let divrem_u n d =
+    if d = Rep.zero then raise Numerics.IntegerDivideByZero else
+      let t = Rep.shift_right d (32 - 1) in
```

Comment:
This 32 - 1 was 64 - 1 in the original i64 code. I don't yet know if this means the algorithm works either way or if we need more tests.